docs/developer-guide/plugins-scenarios.md

```diff
@@ -4,258 +4,342 @@
-the HTTP headers above. The example below transforms the HTTP response content in as streaming fashion using a 
-transformContent() function. The new response is created by invoking a response.newBuilder() method. Using the 
-resulting builder, the response body is overridden. 
+Step 1. We start by calling `chain.proceed(request)` to obtain a response observable. 
 
-	chain.proceed(request)
-	  .map((HttpResponse response) -> response.newBuilder()
-	      .body(response.body().content().map(this::transformContent))
-	      .build()
-	  );
+Step 2. A `callTo3rdParty` returns a CompletableFuture. We use `StyxObservable.from` to convert it 
+into a `StyxObservable` so that it can be bound to the response observable. 
 
-As always, the content can be transformed both synchronously or asynchronously. In this document we will explore all the options.
+Step 3. The `flatMap` operator binds `callToThirdParty` into the response observable.
 
-Content can be decoded to a business domain object for further processing.
+Step 4. We will transform the HTTP response by inserting an outcome of `callToThirdParty`, or `value`, 
+into the response headers.	
 
-### Decode content into business domain objects
 
-Decoding content into a business domain object is always an asynchronous operation. This is because Styx must wait 
-until it has received enough HTTP content to attempt decoding content.
+## HTTP Content Transformations	
 
-HttpResponse object has a decode() method that can be used to decode HTTP response content into business domain objects. 
-The decode() method has the following type:
+Styx exposes HTTP messages to interceptors as streaming `HttpRequest` and `HttpResponse` messages.
+In this form, the interceptors can process the content in a streaming fashion. That is, they they
+can look into, and modify the content as it streams through.
 
-    public <T> Observable<DecodedResponse<T>> decode(Function<ByteBuf, T> decoder, int maxContentBytes)
-    
-The decode() takes two arguments: a decoder function, and a maxContentBytes integer. It returns an Observable<DecodedResponse<T>> 
-which is underlines its asynchronous nature. The decode() accumulates entire HTTP response content into a single aggregated 
-byte buffer, and then calls the supplied  decoder method passing in the aggregated buffer as a parameter. The decoder 
-is a function, or typically a lambda expression supplied by the plugin, that converts the aggregated http content into 
-a business domain object of type T. Styx will take care of managing all aspects related to direct memory buffers, such 
-as releasing them if necessary.
+Alternatively, streaming messages can be aggregated into a `FullHttpRequest` or `FullHttpResponse` 
+messages. The full HTTP message body is then available at interceptor's disposal. Note that content
+aggregation is always an asynchronous operation. This is because Styx must wait until all content
```

Comment:
No. This is because content aggregation will happen asynchronously anyway, and therefore it won't be blocking a thread. 

The real reason why the aggregation has to happen asynchronously is because the content simply won't be available in full at the time of aggregation. Styx simply has to wait until the underlying stream produces the necessary content.

Well, one could say that it "must be performed asynchronously to prevent blocking the IO thread", but that is missing the point, really.

I have attempted to clarify this:

```
Styx exposes HTTP messages to interceptors as streaming `HttpRequest` and `HttpResponse` messages.
In this form, the interceptors can process the content in a streaming fashion. That is, they they
can look into, and modify the content as it streams through.

Alternatively, streaming messages can be aggregated into a `FullHttpRequest` or `FullHttpResponse` 
messages. The full HTTP message body is then available at interceptor's disposal. Note that content
aggregation is always an asynchronous operation. This is because the streaming HTTP message is 
exposing the content, in byte buffers, as it arrives from the network, and Styx must wait until 
all content has been received. 
```

What do you think?